File: app/app/src/main/java/com/cloudcast/app/crypto/UserKeys.kt
package com.cloudcast.app.crypto

import android.util.Base64

object UserKeys {

    class NewUserKeys(

        val salt: ByteArray,

        val publicKey: ByteArray,

        val encryptedPrivateKey: ByteArray,

        val recoveryEncryptedPrivateKey: ByteArray,

        val mkSealed: ByteArray,

        val recoveryCode: String,
    )

    fun createUserKeys(password: String): NewUserKeys {
        val salt = Sodium.generateSalt()
        val kp = Sodium.generateKeypair()
        val recoveryKeyBytes = Sodium.randomBytes(32)
        val recoveryCode = encodeRecoveryCode(recoveryKeyBytes)
        val mk = Sodium.randomBytes(32)
        val vaultKey = Sodium.deriveVaultKey(password, salt)
        val rVaultKey = Sodium.recoveryVaultKey(recoveryKeyBytes)

        val out = NewUserKeys(
            salt = salt,
            publicKey = kp.publicKey,
            encryptedPrivateKey = Sodium.secretboxSeal(kp.privateKey, vaultKey),
            recoveryEncryptedPrivateKey = Sodium.secretboxSeal(kp.privateKey, rVaultKey),
            mkSealed = Sodium.sealTo(mk, kp.publicKey),
            recoveryCode = recoveryCode,
        )

        kp.privateKey.wipe(); vaultKey.wipe(); rVaultKey.wipe(); recoveryKeyBytes.wipe(); mk.wipe()
        return out
    }

    fun openMk(mkSealed: ByteArray, publicKey: ByteArray, privateKey: ByteArray): ByteArray =
        Sodium.sealOpen(mkSealed, publicKey, privateKey)

    fun privateKeyFromPassword(password: String, salt: ByteArray, encryptedPrivateKey: ByteArray): ByteArray {
        val vaultKey = Sodium.deriveVaultKey(password, salt)
        try {
            return Sodium.secretboxOpen(encryptedPrivateKey, vaultKey)
        } finally {
            vaultKey.wipe()
        }
    }

    fun privateKeyFromRecovery(recoveryCode: String, recoveryEncryptedPrivateKey: ByteArray): ByteArray {
        val raw = decodeRecoveryCode(recoveryCode)
        val rVaultKey = Sodium.recoveryVaultKey(raw)
        try {
            return Sodium.secretboxOpen(recoveryEncryptedPrivateKey, rVaultKey)
        } finally {
            raw.wipe(); rVaultKey.wipe()
        }
    }

    fun rewrapForNewPassword(privateKey: ByteArray, newPassword: String, salt: ByteArray): ByteArray {
        val vaultKey = Sodium.deriveVaultKey(newPassword, salt)
        try {
            return Sodium.secretboxSeal(privateKey, vaultKey)
        } finally {
            vaultKey.wipe()
        }
    }

    fun wrapDekFor(dek: ByteArray, recipientPk: ByteArray): ByteArray = Sodium.sealTo(dek, recipientPk)

    fun unwrapDek(wrapped: ByteArray, pk: ByteArray, sk: ByteArray): ByteArray = Sodium.sealOpen(wrapped, pk, sk)

    private const val B64 = Base64.URL_SAFE or Base64.NO_PADDING or Base64.NO_WRAP

    fun encodeRecoveryCode(raw: ByteArray): String = Base64.encodeToString(raw, B64)

    fun decodeRecoveryCode(code: String): ByteArray = Base64.decode(code.trim(), B64)
}

File: web/src/features/crypto/userkeys.ts
import * as cc from "./sodium";

export interface NewUserKeys {
  salt: Uint8Array;
  publicKey: Uint8Array;
  encryptedPrivateKey: Uint8Array;
  recoveryEncryptedPrivateKey: Uint8Array;
  mkSealed: Uint8Array;
  recoveryCode: string;
}

export function createUserKeys(password: string): NewUserKeys {
  const salt = cc.generateSalt();
  const kp = cc.generateKeypair();
  const recoveryKeyBytes = cc.randomBytes(32);
  const recoveryCode = cc.toB64Url(recoveryKeyBytes);
  const mk = cc.randomBytes(32);
  const vaultKey = cc.deriveVaultKey(password, salt);
  const rVaultKey = cc.recoveryVaultKey(recoveryKeyBytes);
  const out: NewUserKeys = {
    salt,
    publicKey: kp.publicKey,
    encryptedPrivateKey: cc.secretboxSeal(kp.privateKey, vaultKey),
    recoveryEncryptedPrivateKey: cc.secretboxSeal(kp.privateKey, rVaultKey),
    mkSealed: cc.sealTo(mk, kp.publicKey),
    recoveryCode,
  };
  cc.wipe(kp.privateKey, vaultKey, rVaultKey, recoveryKeyBytes, mk);
  return out;
}

export function createRecoveryVault(privateKey: Uint8Array): { recoveryCode: string; recoveryEncryptedPrivateKey: Uint8Array } {
  const recoveryKeyBytes = cc.randomBytes(32);
  const out = {
    recoveryCode: cc.toB64Url(recoveryKeyBytes),
    recoveryEncryptedPrivateKey: cc.secretboxSeal(privateKey, cc.recoveryVaultKey(recoveryKeyBytes)),
  };
  cc.wipe(recoveryKeyBytes);
  return out;
}

export const openMk = (mkSealed: Uint8Array, pk: Uint8Array, sk: Uint8Array): Uint8Array => cc.sealOpen(mkSealed, pk, sk);

export const privateKeyFromPassword = (password: string, salt: Uint8Array, enc: Uint8Array): Uint8Array =>
  cc.secretboxOpen(enc, cc.deriveVaultKey(password, salt));

export const privateKeyFromRecovery = (recoveryCode: string, enc: Uint8Array): Uint8Array =>
  cc.secretboxOpen(enc, cc.recoveryVaultKey(cc.fromB64Url(recoveryCode)));

export const rewrapForNewPassword = (privateKey: Uint8Array, newPassword: string, salt: Uint8Array): Uint8Array =>
  cc.secretboxSeal(privateKey, cc.deriveVaultKey(newPassword, salt));

export const wrapDekFor = (dek: Uint8Array, recipientPk: Uint8Array): Uint8Array => cc.sealTo(dek, recipientPk);
export const unwrapDek = (wrapped: Uint8Array, pk: Uint8Array, sk: Uint8Array): Uint8Array => cc.sealOpen(wrapped, pk, sk);
